cmd/geth/chaincmd.go

```diff
@@ -675,8 +711,202 @@ func dump(ctx *cli.Context) error {
 	return nil
 }
 
+func exportSegment(ctx *cli.Context) error {
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	db := utils.MakeChainDatabase(ctx, stack, false, false)
+	defer db.Close()
+
+	genesisHash := rawdb.ReadCanonicalHash(db, 0)
+	td := rawdb.ReadTd(db, genesisHash, 0)
+	chainConfig, engine, headerChain, err := simpleHeaderChain(db, genesisHash)
+	if err != nil {
+		return err
+	}
+	latest := headerChain.CurrentHeader()
+	if _, ok := engine.(consensus.PoSA); !ok {
+		return errors.New("cannot generate history segment because consensus engine is not PoSA")
+	}
+	if !chainConfig.IsLuban(latest.Number) {
+		return errors.New("luban hard fork is not enabled , cannot generate history segment")
+	}
+
+	var (
+		boundStartBlock      = params.BoundStartBlock
+		historySegmentLength = params.HistorySegmentLength
+	)
+	switch genesisHash {
+	case params.BSCGenesisHash, params.ChapelGenesisHash, params.RialtoGenesisHash:
+		boundStartBlock = params.BoundStartBlock
+		historySegmentLength = params.HistorySegmentLength
+	default:
+		if ctx.IsSet(utils.BoundStartBlockFlag.Name) {
+			boundStartBlock = ctx.Uint64(utils.BoundStartBlockFlag.Name)
+		}
+		if ctx.IsSet(utils.HistorySegmentLengthFlag.Name) {
+			historySegmentLength = ctx.Uint64(utils.HistorySegmentLengthFlag.Name)
+		}
+	}
+	if boundStartBlock == 0 || historySegmentLength == 0 {
+		return fmt.Errorf("wrong params, boundStartBlock: %v, historySegmentLength: %v", boundStartBlock, historySegmentLength)
+	}
+
+	latestNum := latest.Number.Uint64()
+	if latestNum < params.FullImmutabilityThreshold || latestNum < boundStartBlock {
+		return errors.New("current chain is too short, less than BoundStartBlock")
+	}
+
+	start := time.Now()
+	target := latestNum - params.FullImmutabilityThreshold
+	log.Info("start export segment", "from", boundStartBlock, "to", target, "boundStartBlock", boundStartBlock,
+		"historySegmentLength", historySegmentLength, "chainCfg", chainConfig)
+	segments := []params.HistorySegment{
+		{
+			Index:           0,
+			ReGenesisNumber: 0,
+			ReGenesisHash:   genesisHash,
+			TD:              td.Uint64(),
+		},
+	}
+	// try find finalized block in every segment boundary
+	for num := boundStartBlock; num <= target; num += historySegmentLength {
+		// align the segment start at parlia's epoch
```

Comment:
why epoch aligned?

Reply:
When parlia verifies the header, it needs a snapshot produced from the epoch block header. So it is better to align the segment starting at parlia's epoch, the parlia always could find that info.